internal/utils/sign.go

```diff
@@ -94,26 +109,66 @@ func (e *Signer) SignWithKey(key crypto.PrivateKey, msg []byte) ([]byte, error)
 		if err != nil {
 			return nil, err
 		}
+	case rsa.PrivateKey:
+		if e.signatureType == SignatureTypeRSAPKCS1 {
+			opts := crypto.SignerOpts(e.hash)
+			sig, err = sk.Sign(e.rand, msg, opts)
+		} else {
+			opts := &rsa.PSSOptions{Hash: e.hash}
+			sig, err = sk.Sign(e.rand, msg, opts)
+		}
+		if err != nil {
+			return nil, err
+		}
 	default:
-		return nil, fmt.Errorf("tls: unsupported key type")
+		return nil, fmt.Errorf("unsupported key type")
 	}
 
 	return sig, nil
 }
 
 // TODO(claucece): as this is used beyond DCs, it needs to support all the other algos.
-func getSigner(bugs *CertificateBugs, rand io.Reader, sigAlg uint16) (*Signer, error) {
+func getSigner(randReader io.Reader, sigAlg uint16, selectValidDCAlg bool) (*Signer, error) {
+	if sigAlg == 0 { // Choose one at random
+		sigAlgList := []uint16{SignatureECDSAWithP256AndSHA256,
+			SignatureECDSAWithP384AndSHA384,
+			SignatureECDSAWithP521AndSHA512,
+			SignatureEd25519}
+		if !selectValidDCAlg {
+			sigAlgList = append(sigAlgList,
+				[]uint16{SignatureRSAPKCS1WithMD5,
+					SignatureRSAPKCS1WithSHA1,
+					SignatureRSAPKCS1WithSHA256,
+					SignatureRSAPKCS1WithSHA384,
+					SignatureRSAPKCS1WithSHA512}...)
+		}
+
+		buf := make([]byte, 1)
+		_, err := randReader.Read(buf)
+		if err != nil {
+			return nil, err
+		}
+		sigAlg = sigAlgList[int(buf[0])%len(sigAlgList)]
+	}
 	switch sigAlg {
-	case SignatureECDSAWithSHA1:
-		return &Signer{bugs, nil, crypto.SHA1, nil, rand, true}, nil
 	case SignatureECDSAWithP256AndSHA256:
-		return &Signer{bugs, elliptic.P256(), crypto.SHA256, nil, rand, true}, nil
+		return &Signer{SignatureECDSAWithP256AndSHA256, SignatureTypeECDSA, randReader, crypto.SHA256, nil, elliptic.P256(), 0}, nil
 	case SignatureECDSAWithP384AndSHA384:
-		return &Signer{bugs, elliptic.P384(), crypto.SHA384, nil, rand, true}, nil
+		return &Signer{SignatureECDSAWithP384AndSHA384, SignatureTypeECDSA, randReader, crypto.SHA384, nil, elliptic.P384(), 0}, nil
 	case SignatureECDSAWithP521AndSHA512:
-		return &Signer{bugs, elliptic.P521(), crypto.SHA512, nil, rand, true}, nil
+		return &Signer{SignatureECDSAWithP521AndSHA512, SignatureTypeECDSA, randReader, crypto.SHA512, nil, elliptic.P521(), 0}, nil
 	case SignatureEd25519:
-		return &Signer{bugs, nil, directSigning, nil, rand, false}, nil
+		return &Signer{SignatureEd25519, SignatureTypeEd25519, randReader, directSigning, nil, nil, 0}, nil
+	case SignatureRSAPKCS1WithMD5:
+		return &Signer{SignatureRSAPKCS1WithMD5, SignatureTypeRSAPKCS1, randReader, crypto.MD5, nil, nil, 2048}, nil
+	case SignatureRSAPKCS1WithSHA1:
+		return &Signer{SignatureRSAPKCS1WithSHA1, SignatureTypeRSAPKCS1, randReader, crypto.SHA1, nil, nil, 2048}, nil
+	case SignatureRSAPKCS1WithSHA256:
+		return &Signer{SignatureRSAPKCS1WithSHA256, SignatureTypeRSAPKCS1, randReader, crypto.SHA1, nil, nil, 2048}, nil
```

Comment:
If we are going to use them, I think we need to change the sha.

Reply:
Ah, this was a typo, will fix.

Reply:
Thank you! Otherwise, looking super nicely! Let's sync tomorrow ;)